Adding HAVOK to PyDMD

sichinaga opened the pull request:
Hello again to everyone on the PyDMD team!

I am here to propose the next potential addition to the PyDMD family: the Hankel Alternative View of Koopman (HAVOK) analysis! In case anyone is interested, the article deriving and describing the method can be found here: https://www.nature.com/articles/s41467-017-00030-8

HAVOK is a method specifically tailored for modeling one-dimensional time-series datasets that exhibit chaos. In short, the algorithm can be broken down into the following steps:

1) Compute a hankel matrix $H$ from the input time-series.
2) Compute the SVD of the hankel matrix $H = U\Sigma V^*$.
3) Apply DMD to $V^T$ and obtain the continuous-time DMD operator $A_c$ in $\dot{V}^T = A_c V^T$.
4) Define $A$ as the matrix $A_c$ but excluing the last row and column. Define $B$ as the last column of $A_c$, but excluding the last entry.
5) We may then model our data via the relationship $\dot{v} = Av + Bv_r$, where $V = [v_1 \quad \dots \quad v_r]$ and $v = [v_1 \quad \dots \quad v_{r-1}]^T$.

Given the structure of the HAVOK analysis, I have made the following implementation choices:
- The `HAVOK` class is a child of the `HankelDMD` class due to their shared usage of hankel matrices. Note that the key differences between the two methods are (1) HAVOK applies DMD to the $V$ matrix from the SVD of $H$ while Hankel DMD applies DMD to $H$, and (2) HAVOK utilizes a forcing term in order to reconstruct data, while Hankel DMD does not.
- In order to utilize currently-existing DMD modules, I compute the continuous-time operator $A_c$ by first using DMD to compute the discrete-time operator $A_d$ in $V_2^T = A_d V_1^T$. I then utilize the approximation $A_c \approx (A_d - I) / \Delta t$. There are most certainly better/more accurate methods of computing $A_c$, but I figured this rough approximation should suffice for now.
- I'm currently using `scipy.signal` in order to compute the HAVOK reconstructions (since this is how reconstructions are computed in the original HAVOK paper, and the code is quite concise), but I could also envision this being implemented via the `DMDc` module somehow... I'm happy to keep the reconstruction code as is, but if reusing code from `DMDc` has the potential to be stylistically better, I'd be happy to give the implementation a shot!
- One other thing, I've added a `tikhonov_regularization` input parameter to `HankelDMD` so that it may now be used by both `HankelDMD` and `HAVOK`. Let me know if this parameter was omitted from `HankelDMD` for any particular reason so that I can restore `HankelDMD` to its original form! :)

Thank you all again for your diligence and support with these new modules. 😄 

I'm also open to any feedback, comments, questions, or concerns that may arise with this module. :)

fAndreuzzi commented on pydmd/havok.py:
```diff
@@ -0,0 +1,194 @@
+        :return: matrix containing the linear time-delay embeddings.
+        :rtype: numpy.ndarray
+        """
+        return self._embeddings[:,:-1]
```
This will trigger a weird error if `fit()` is not called

```suggestion
        if self._embeddings is None:
                raise RuntimeError('fit() not called')
        return self._embeddings[:,:-1]
```

fAndreuzzi commented on pydmd/havok.py:
```diff
@@ -0,0 +1,194 @@
+        :return: array containing the chaotic forcing term.
+        :rtype: numpy.ndarray
+        """
+        return self._embeddings[:,-1]
```
Same as `linear_embeddings`

sichinaga replied:
Unless I am being silly, `linear_embeddings` and `forcing_input` are slightly different in that `linear_embeddings` returns all of the columns of `self._embeddings` excluding the last column, while `forcing_input` returns only the last column of `self._embeddings`.

fAndreuzzi replied:
Sorry @sichinaga my fault, I meant that my comment here is the same I wrote for `linear_embeddings`:

`This will trigger a weird error if fit() is not called`

sichinaga replied:
Ahhh gotcha. Thank you for clarifying! :)

fAndreuzzi commented on pydmd/havok.py:
```diff
@@ -0,0 +1,194 @@
+        :return: linear dynamics matrix A.
+        :rtype: numpy.ndarray
+        """
+        return self._A
```
I believe it's good to throw an exception here too, like for `linear_embeddings`

fAndreuzzi commented on pydmd/havok.py:
```diff
@@ -0,0 +1,194 @@
+        :return: forcing dynamics vector B.
+        :rtype: numpy.ndarray
+        """
+        return self._B
```
Same as `A` above

fAndreuzzi commented on pydmd/havok.py:
```diff
@@ -0,0 +1,194 @@
+            np.diag(self._svd_amps[:-1]),
+            reconstructed_v.conj().T
+        ])
+        reconstructed_x = self._dehankel(reconstructed_hankel_matrix)
+
+        return reconstructed_x
```
I propose to merge these just to make it more compact

```suggestion
        return self._dehankel(reconstructed_hankel_matrix)
```

fAndreuzzi commented on pydmd/havok.py:
```diff
@@ -0,0 +1,194 @@
+        self._snapshots = self._snapshots.squeeze()
+
+        # Check that input data is a 1D time-series
+        if np.ndim(self._snapshots) != 1:
```
It's the same but it's more readable:
```suggestion
        if self._snapshots.ndim > 1:
```

fAndreuzzi commented on pydmd/havok.py:
```diff
@@ -0,0 +1,194 @@
+            raise ValueError(msg.format(self._d))
+
+        # Compute hankel matrix for the input data
+        self._hankel_matrix = self._pseudo_hankel_matrix(
+            self._snapshots.reshape(1, -1)
+        )
```
It's the same but it's more readable:
```suggestion
        self._hankel_matrix = self._pseudo_hankel_matrix(self._snapshots[None])
```

fAndreuzzi commented on pydmd/havok.py:
```diff
@@ -0,0 +1,194 @@
+
+        # Save A matrix and B vector 
+        self._A = regression_model_continuous[:-1, :-1]
+        self._B = regression_model_continuous[:-1, -1].reshape(-1, 1)
```
It's the same but it's more readable:
```suggestion
        self._B = regression_model_continuous[:-1, -1, None]
```

fAndreuzzi reviewed (commented):
Hi @sichinaga, thank you for the new DMD. I left some minor comments on the code. It looks very good, do you mind adding a few general tests, just to make sure it can run and produce reasonable results?

sichinaga commented:
Thank you @fAndreuzzi for all of your great feedback so far! 😃

I will begin working on some tests now. The commit should be up soon. :)

sichinaga commented:
Hi again @fAndreuzzi!

Just dropping in to let you know that I have now uploaded a small suite of tests for the `HAVOK` module! I've also added in a few small edits including a new attribute and some extra precautionary checks. The edits are overall quite minor though. 

With that being said, the test suite that I uploaded is arguably quite small in comparison to the other PyDMD test suites, but I hope it serves as a good starting point! Each test is documented in case you're curious about the breadth of the current batch of tests. :) I'm happy to add many more tests, especially if there are any major ones that I left out. I just implemented the ones that I felt were most relevant to the HAVOK algorithm, but please let me know if there's anything else you feel I can do to improve the quality of this PR.

Thanks again for your kind support. :)

ndem0 commented:
Dear @sichinaga, thanks for the addition! @fAndreuzzi already did the review, so everything is fine!

I'm going to merge this PR, we can always add new tests in future, but these are enough to avoid critical issues in the development.

fAndreuzzi commented:
Hi @sichinaga, thanks for adding the tests, it looks like we have a very good coverage :)
Good job!
